symengine/sets.cpp

```diff
@@ -464,4 +554,250 @@ RCP<const Boolean> Union::contains(const RCP<const Basic> &o) const
+    return container_->set_complement(newuniv);
+}
+
+RCP<const Set> set_union(const set_set &in, bool solve)
+{
+    set_set input;
+    if (solve == false && in.size() > 1)
+        return make_rcp<const Union>(in);
+    set_basic combined_FiniteSet;
+    for (auto it = in.begin(); it != in.end(); ++it) {
+        if (is_a<FiniteSet>(**it)) {
+            const FiniteSet &other = down_cast<const FiniteSet &>(**it);
+            combined_FiniteSet.insert(other.get_container().begin(),
+                                      other.get_container().end());
+        } else if (is_a<UniversalSet>(**it)) {
+            return universalset();
+        } else if (not is_a<EmptySet>(**it)) {
+            input.insert(*it);
+        }
+    }
+    if (input.empty()) {
+        return finiteset(combined_FiniteSet);
+    } else if (input.size() == 1 && combined_FiniteSet.empty()) {
+        return *input.begin();
+    }
+    // Now we rely on respective containers' own rules
+    // TODO: Improve it to O(log n)
+    RCP<const Set> combined_Rest = finiteset(combined_FiniteSet);
+    for (auto it = input.begin(); it != input.end(); ++it) {
+        combined_Rest = combined_Rest->set_union(*it);
+    }
+    return combined_Rest;
+}
+
+RCP<const Set> set_intersection(const set_set &in, bool solve)
+{
+    if (solve == false)
+        throw std::runtime_error(
+            "Not implemented"); // No intersection class yet
+
+    // https://en.wikipedia.org/wiki/Intersection_(set_theory)#Nullary_intersection
+    if (in.empty())
+        return universalset();
+
+    // Global rules
+    // If found any emptyset then return emptyset
+    set_set incopy;
+    for (const auto &input : in) {
+        if (is_a<EmptySet>(*input))
+            return emptyset();
+        else if (not is_a<UniversalSet>(*input))
+            incopy.insert(input);
+    }
+
+    if (incopy.empty())
+        return universalset();
+    if (incopy.size() == 1)
+        return *incopy.begin();
+
+    // Handle finite sets
+    set_set fsets, othersets;
+    for (const auto &input : incopy) {
+        if (is_a<FiniteSet>(*input))
+            fsets.insert(input);
+        else
+            othersets.insert(input);
+    }
+    if (fsets.size() != 0) {
+        const FiniteSet &fs = down_cast<const FiniteSet &>(**fsets.begin());
+        auto cont = fs.get_container();
+        fsets.erase(fsets.begin());
+        set_basic finalfs;
+        for (const auto &fselement : cont) {
+            bool present = true;
+            for (const auto &fset : fsets) {
+                present
+                    = present and eq(*(fset->contains(fselement)), *boolTrue);
+            }
+            if (!present)
+                continue;
+            for (const auto &oset : othersets) {
+                present
+                    = present and eq(*(oset->contains(fselement)), *boolTrue);
+            }
+            if (present)
+                finalfs.insert(fselement);
```

Comment:
It should return runtime error as we don't have a class to represent its Intersection.
earlier this code gave `emptyset()` . but i fixed it now.